File: src/Scanner.Core/Checks/MissingDocumentLanguageCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Detects missing language declarations on HTML documents.
/// </summary>
public sealed class MissingDocumentLanguageCheck : ICheck
{
    private static readonly Regex HtmlTagRegex = new("<\\s*html(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "missing-document-language";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var match = HtmlTagRegex.Match(context.Content);
        if (!match.Success)
        {
            yield break;
        }

        var attrs = match.Groups["attrs"].Value;
        var lang = AttributeParser.GetAttributeValue(attrs, "lang");
        var xmlLang = AttributeParser.GetAttributeValue(attrs, "xml:lang");

        if (!string.IsNullOrWhiteSpace(lang) || !string.IsNullOrWhiteSpace(xmlLang))
        {
            yield break;
        }

        var line = TextUtilities.GetLineNumber(context.Content, match.Index);
        yield return new Issue(rule.Id, Id, context.FilePath, line, "Document language is missing or empty.", match.Value);
    }
}

File: node_engine/src/checks/MissingDocumentLanguageCheck.js
const { getAttributeValue } = require("./AttributeParser");
const { getLineNumber, getLineNumberForSnippet } = require("./TextUtilities");

const htmlTagRegex = /<\s*html(?<attrs>[^>]*)>/i;

const MissingDocumentLanguageCheck = {
  id: "missing-document-language",
  applicableKinds: ["html", "htm"],
  run(context, rule) {
    if (context.document?.documentElement) {
      const root = context.document.documentElement;
      const lang = root.getAttribute("lang");
      const xmlLang = root.getAttribute("xml:lang");
      if ((lang && lang.trim()) || (xmlLang && xmlLang.trim())) {
        return [];
      }
      const evidence = root.outerHTML;
      return [
        {
          ruleId: rule.id,
          checkId: MissingDocumentLanguageCheck.id,
          filePath: context.filePath,
          line: getLineNumberForSnippet(context.content, evidence),
          message: "Document language is missing or empty.",
          evidence
        }
      ];
    }

    const match = context.content.match(htmlTagRegex);
    if (!match) {
      return [];
    }

    const attrs = match.groups.attrs;
    const lang = getAttributeValue(attrs, "lang");
    const xmlLang = getAttributeValue(attrs, "xml:lang");

    if ((lang && lang.trim()) || (xmlLang && xmlLang.trim())) {
      return [];
    }

    const line = getLineNumber(context.content, match.index);
    return [
      {
        ruleId: rule.id,
        checkId: MissingDocumentLanguageCheck.id,
        filePath: context.filePath,
        line,
        message: "Document language is missing or empty.",
        evidence: match[0]
      }
    ];
  }
};

module.exports = { MissingDocumentLanguageCheck };
